pkg/store/bucket.go

```diff
@@ -2339,62 +2414,43 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, offs []uint32, seq i
 		fetchBegin = time.Now()
 
 		// Read entire chunk into new buffer.
-		nb, err := r.block.readChunkRange(ctx, seq, int64(o), int64(chLen), []byteRange{{offset: 0, length: chLen}})
+		nb, err := r.block.readChunkRange(ctx, seq, int64(pIdx.off), int64(chunkLen), []byteRange{{offset: 0, length: chunkLen}})
 		if err != nil {
-			return errors.Wrapf(err, "preloaded chunk too small, expecting %d, and failed to fetch full chunk", chLen)
+			return errors.Wrapf(err, "preloaded chunk too small, expecting %d, and failed to fetch full chunk", chunkLen)
 		}
-
-		cb = *nb
-		if len(cb) != chLen {
-			return errors.Errorf("preloaded chunk too small, expecting %d", chLen)
+		if len(*nb) != chunkLen {
+			return errors.Errorf("preloaded chunk too small, expecting %d", chunkLen)
 		}
 
 		r.mtx.Lock()
 		locked = true
 
-		r.chunkBytes = append(r.chunkBytes, nb)
 		r.stats.chunksFetchCount++
 		r.stats.chunksFetchDurationSum += time.Since(fetchBegin)
-		r.stats.chunksFetchedSizeSum += len(cb)
+		r.stats.chunksFetchedSizeSum += len(*nb)
 
-		r.chunks[chunkRef] = rawChunk(cb[n:])
-	}
-	return nil
-}
-
-// chunkOffsetsToByteRanges returns non-overlapping byte ranges with each range offset
-// relative to start. The provided input offsets must be sorted.
-func chunkOffsetsToByteRanges(offsets []uint32, start uint32) byteRanges {
-	ranges := make([]byteRange, len(offsets))
-
-	for idx := 0; idx < len(offsets); idx++ {
-		ranges[idx] = byteRange{
-			// The byte range offset is required to be relative to the start of the read slice.
-			offset: int(offsets[idx] - start),
-			length: EstimatedMaxChunkSize,
+		if err := populateChunk(&(res[pIdx.i].chks[pIdx.j]), rawChunk((*nb)[n:]), aggrs, r.savior); err != nil {
+			return errors.Wrap(err, "populate chunk")
 		}
 
-		if idx > 0 {
-			// Ensure ranges are non overlapping.
-			if prev := ranges[idx-1]; prev.length > ranges[idx].offset-prev.offset {
-				ranges[idx-1].length = ranges[idx].offset - prev.offset
-			}
-		}
+		r.block.chunkPool.Put(nb)
 	}
-
-	return ranges
+	return nil
 }
 
-func (r *bucketChunkReader) Chunk(id uint64) (chunkenc.Chunk, error) {
-	c, ok := r.chunks[id]
-	if !ok {
-		return nil, errors.Errorf("chunk with ID %d not found", id)
+func (r *bucketChunkReader) savior(b []byte) ([]byte, error) {
```

Comment:
yes, there is a way to optimize this further - as currently slab "tails" are wasted. This would potentially save up to `max(raw values chunk / subchunk size)` - 1 per slab (set to 64k in this PR). What is that, 3200 bytes? If so, maximum gain is 4.9% per slab.

However, for the benchmarks we have (filled with random values), average size tends to be less than half of that, so that would cut our estimated maximum profit to, say, 1500 bytes per slab (2.3%). Given that not every case would be that bad, I would estimate actual memory wastes to be around 1.5%.

Moreover, we would need to be lucky and only write chunks that are small enough to be written to the wasted space - so we would not be able to reclaim all of it. All in all, I would consider it a win if we reduce that waste from 1.5% to 1%. So, 5MB save on a 1GB allocation. Do we have bigger fish to fry?

Reply:
Makes sense, also - one thing at the time (: